contracts/PrepaidCardMarket_v2.sol

```diff
@@ -0,0 +1,304 @@
+    address issuerSafe = skus[sku].issuerSafe;
+
+    IERC677(token).safeTransferAndCall(
+      prepaidCardManagerAddress,
+      priceToCreatePrepaidCard,
+      abi.encode(
+        customer,
+        issuingTokenAmounts,
+        spendAmounts,
+        customizationDID,
+        address(0), // marketAddress - we probably don't need it in this case?
+        issuer,
+        issuerSafe
+      )
+    );
+
+    balance[issuerSafe][token] -= priceToCreatePrepaidCard;
+
+    return true;
+  }
+
+  function getQuantity(bytes32 sku) public view returns (uint256) {
+    PrepaidCardManager prepaidCardManager = PrepaidCardManager(
+      prepaidCardManagerAddress
+    );
+
+    address token = skus[sku].issuingToken;
+    uint256 faceValue = skus[sku].faceValue;
+    address issuerSafe = skus[sku].issuerSafe;
+
+    uint256 price = prepaidCardManager.priceForFaceValue(token, faceValue);
+
+    return balance[issuerSafe][token] / price;
+  }
+
+  function setAsk(
+    address issuer,
+    bytes32 sku,
+    uint256 askPrice
+  ) external returns (bool) {
+    require(skus[sku].issuer != address(0), "Non-existent SKU");
+    require(skus[sku].issuer == issuer, "SKU not owned by issuer");
+    asks[sku] = askPrice;
+
+    emit AskSet(issuer, skus[sku].issuingToken, sku, askPrice);
+    return true;
+  }
+
+  function addSKU(
+    uint256 faceValue,
+    string memory customizationDID,
+    address token
+  ) external returns (bool) {
+    require(faceValue > 0, "Face value must be greater than 0");
+    require(token != address(0), "Token address must be set");
+
+    address _issuer = issuers[msg.sender];
+    require(_issuer != address(0), "Issuer not found.");
+
+    bytes32 sku = getSKU(_issuer, token, faceValue, customizationDID);
+
+    require(skus[sku].issuer == address(0), "SKU already exists");
+
+    skus[sku] = SKU({
+      issuerSafe: msg.sender,
+      issuer: _issuer,
+      issuingToken: token,
+      faceValue: faceValue,
+      customizationDID: customizationDID
+    });
+
+    emit SkuAdded(_issuer, token, faceValue, customizationDID, sku);
+
+    return true;
+  }
+
+  function getSKU(
+    address issuer,
+    address token,
+    uint256 faceValue,
+    string memory customizationDID
+  ) public pure returns (bytes32) {
+    return
+      keccak256(abi.encodePacked(issuer, token, faceValue, customizationDID));
+  }
+
+  function withdrawTokens(uint256 amount, address token) external {
+    address _issuer = issuers[msg.sender];
+    require(_issuer != address(0), "Issuer not found");
+
+    uint256 balanceForToken = balance[msg.sender][token];
+
+    require(amount <= balanceForToken, "Insufficient funds for withdrawal");
+
+    balance[msg.sender][token] -= amount;
+
+    IERC677(token).safeTransfer(msg.sender, amount);
+
+    emit TokensWithdrawn(msg.sender, _issuer, token, amount);
+  }
+
+  function onTokenTransfer(
+    address payable from, // safe address
+    uint256 amount,
+    bytes calldata data
```

Comment:
let's document what the `calldata` is (encoded address specifying the issuer) to make it easy for folks reading the code

Reply:
Added in https://github.com/cardstack/card-pay-protocol/pull/191/commits/7f6f29ca8720176bf2569a4326b8e89d9b961f6f